service/worker/statsreporter/statsreporter.go

```diff
@@ -0,0 +1,164 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package statsreporter
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/membership"
+	"go.temporal.io/server/common/metrics"
+	ns "go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/internal/goro"
+)
+
+const listNamespacePageSize = 100
+
+type StatsReporter struct {
+	status                int32
+	reporter              *goro.Handle
+	visibilityRateLimiter quotas.RateLimiter
+
+	Logger                log.Logger
+	MetricsHandler        metrics.MetricsHandler
+	MetadataManager       persistence.MetadataManager
+	VisibilityManager     manager.VisibilityManager
+	BaseReportInterval    dynamicconfig.DurationPropertyFn
+	EmitOpenWorkflowCount dynamicconfig.BoolPropertyFnWithNamespaceFilter
+	CountWorkflowMaxQPS   dynamicconfig.IntPropertyFn
+	ServiceResolver       membership.ServiceResolver
+}
+
+// Start is called to start replicator
+func (r *StatsReporter) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&r.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	r.reporter = goro.NewHandle(context.Background()).Go(r.queryLoop)
+	r.visibilityRateLimiter = quotas.NewDefaultOutgoingRateLimiter(
+		func() float64 { return float64(r.CountWorkflowMaxQPS()) },
+	)
+	r.Logger.Info("Stats reporter started.")
+}
+
+// Stop is called to stop replicator
+func (r *StatsReporter) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&r.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	r.reporter.Cancel()
+	<-r.reporter.Done()
+}
+
+func (r *StatsReporter) queryLoop(ctx context.Context) error {
+	timer := time.NewTicker(r.waitDurationTillNextReport())
+
+	for {
+		select {
+		case <-timer.C:
+			r.reportNamespaceStats(ctx)
+			timer.Reset(r.waitDurationTillNextReport())
+		case <-ctx.Done():
+			timer.Stop()
+			return nil
+		}
+	}
+}
+
+func (r *StatsReporter) waitDurationTillNextReport() time.Duration {
+	// BaseReportInterval is the expectation of wait across all workers.
+	// BaseReportInterval*2 is the max wait across workers.
+	// BaseReportInterval*2*worker_count is the total time each worker should wait.
+	// Then we pick a random number within the total time to distribute the load evenly.
+	totalWait := r.BaseReportInterval() * time.Duration(r.ServiceResolver.MemberCount()) * 2
+	return time.Duration(math.Round(float64(totalWait) * rand.Float64()))
+}
+
+func (r *StatsReporter) reportNamespaceStats(ctx context.Context) {
+	var nextPageToken []byte
+	for {
```

Comment:
Use NewPagingIterator?

Reply:
Maybe we should expose getAllNamespace in namespace registry and get all namespace from cache.